gateway/src/apicast/policy/rate_limiting_to_service/rate_limiting_to_service.lua

```diff
@@ -0,0 +1,201 @@
+local next = next
+local shdict_key = 'limitter'
+
+local new = _M.new
+
+function _M.new(config)
+  local self = new()
+  self.config = config or {}
+  self.limitters = config.limitters
+  self.redis_info = config.redis_info
+  return self
+end
+
+local function redis_shdict(host, port, db)
+  local redis = assert(resty_redis:new())
+
+  local ok, connerr = redis:connect(host or '127.0.0.1', port or 6379)
+  if not ok then
+    return nil, connerr
+  end
+
+  ok = redis:select(db or 0)
+  if not ok then
+    return nil, "failed to select db"
+  end
+
+  return {
+    incr = function(_, key, value, init)
+      if not init then
+        return redis:incrby(key, value), nil
+      end
+      redis:setnx(key, init)
+      return redis:incrby(key, value), nil
+    end,
+    set = function(_, key, value)
+      return redis:set(key, value)
+    end,
+    expire = function(_, key, exptime)
+      local ret = redis:expire(key, exptime)
+      if ret == 0 then
+        return nil, "not found"
+      end
+      return true, nil
+    end,
+    get = function(_, key)
+      local val = redis:get(key)
+      if type(val) == "userdata" then
+        return nil
+      end
+      return val
+    end
+  }
+end
+
+local function try(f, catch_f)
+  local status, exception = pcall(f)
+  if not status then
+    catch_f(exception)
+  end
+end
+
+function _M:access()
+  local limitters = {}
+  local keys = {}
+  local states = {}
+
+  local limitters_limit_conn = {}
+  local keys_limit_conn = {}
+
+  local limitters_limit_conn_committed = {}
+  local keys_limit_conn_committed = {}
+
+  if not self.redis_info then
+    ngx.log(ngx.ERR, "No Redis information.")
+    return ngx.exit(500)
+  end
+
+  for _, limitter in ipairs(self.limitters) do
+    local limit
+    local class_not_found = false
+    try(
+      function()
+        limit = require (limitter.limitter)
+      end,
+      function(e)
+        ngx.log(ngx.ERR, "failed to find module: ", e)
+        class_not_found = true
+      end
+    )
+    if class_not_found then
+      return ngx.exit(500)
+    end
+
+    local lim, limerr
+    local failed_to_instantiate = false
+    try(
+      function()
+        lim, limerr = limit.new(shdict_key, unpack(limitter.values))
+        if not lim then
+          ngx.log(ngx.ERR, "failed to instantiate limitter: ", limerr)
+          failed_to_instantiate = true
+        end
+      end,
+      function(e)
+        ngx.log(ngx.ERR, "failed to instantiate limitter: ", e)
+        failed_to_instantiate = true
+      end
+    )
+    if failed_to_instantiate then
+      return ngx.exit(500)
+    end
+
+    local rediserr
+    lim.dict, rediserr = redis_shdict(self.redis_info.host, self.redis_info.port, self.redis_info.db)
+    if not lim.dict then
+      ngx.log(ngx.ERR, "failed to connect Redis: ", rediserr)
+      return ngx.exit(500)
+    end
+
+    limitters[#limitters + 1] = lim
+    keys[#keys + 1] = limitter.key
+
+    if limitter.limitter == "resty.limit.conn" then
```

Comment:
I think this could be abstracted by checking the the limiter has `is_commited` key.